core/ledger/ledgermgmt/ledger_mgmt.go

```diff
@@ -112,10 +123,40 @@ func (m *LedgerMgr) CreateLedger(id string, genesisBlock *common.Block) (ledger.
 	}, nil
 }
 
-// CreateLedgerFromSnapshot creates a new ledger with the given snapshot and returns the ledger and channel id.
-// This function guarantees that the creation of ledger and all ledger dbs would be an atomic action.
-// The channel id retrieved from the snapshot metadata is treated as a ledger id
-func (m *LedgerMgr) CreateLedgerFromSnapshot(snapshotDir string) (ledger.PeerLedger, string, error) {
+// CreateLedgerFromSnapshot creates a new ledger with the given snapshot and executes the callback function
+// after the ledger is created. This function launches to goroutine to create the ledger and call the callback func.
+// All ledger dbs would be created in an atomic action. The channel id retrieved from the snapshot metadata
+// is treated as a ledger id. It returns an error if another ledger is being created from a snapshot.
+func (m *LedgerMgr) CreateLedgerFromSnapshot(snapshotDir string, channelCallback func(ledger.PeerLedger, string)) error {
+	// verify snapshotDir exists and is not empty
+	empty, err := fileutil.DirEmpty(snapshotDir)
+	if err != nil {
+		return err
+	}
+	if empty {
+		return errors.Errorf("snapshot dir %s is empty", snapshotDir)
+	}
+
+	if err := m.setCreateFromSnapshotInfo(snapshotDir); err != nil {
+		return err
+	}
+
+	go func() {
+		defer m.unsetCreateFromSnapshotInfo()
+
+		ledger, cid, err := m.createFromSnapshot(snapshotDir)
+		if err != nil {
+			logger.Errorw("Error creating ledger from snapshot", "snapshotDir", snapshotDir, "error", err)
+			return
+		}
+
+		channelCallback(ledger, cid)
```

Comment:
I guess that it's important to cover this in the test - i.e., on error, callback function is not invoked. It should be easy to cause this error condition in a test by deleting one of the files in the snapshot folder.

Reply:
Added the test.